pkgdb/src/pkgdb/write.cc

```diff
@@ -7,22 +7,293 @@
+      traceLog( "splitting system glob into real systems" );
+      for ( const auto & system : getDefaultSystems() )
+        {
+          AttrPathGlob relPathCopy = relPath;
+          relPathCopy.front()      = system;
+          this->addRule( relPathCopy, rule );
+        }
+      return;
+    }
+
+  std::string attrName = std::move( *relPath.front() );
+  // TODO: Use a `std::deque' instead of `std::vector' for efficiency.
+  //       This container is only used for `push_back' and `pop_front'.
+  //       Removing from the front is inefficient for `std::vector'.
+  relPath.erase( relPath.begin() );
+
+  if ( auto it = this->children.find( attrName ); it != this->children.end() )
+    {
+      traceLog( "found existing child `" + attrName + '\'' );
+      /* Add to existing child node. */
+      it->second.addRule( relPath, rule );
+    }
+  else if ( relPath.empty() )
+    {
+      /* Add leaf node. */
+      traceLog( "creating leaf `" + attrName + " = "
+                + scrapeRuleToString( rule ) + '\'' );
+      this->children.emplace( attrName, RulesTreeNode( attrName, rule ) );
+    }
+  else
+    {
+      traceLog( "creating child `" + attrName + '\'' );
+      /* Create a new child node. */
+      this->children.emplace( attrName, RulesTreeNode( attrName ) );
+      this->children.at( attrName ).addRule( relPath, rule );
+    }
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+ScrapeRule
+RulesTreeNode::getRule( const AttrPath & path ) const
+{
+  const RulesTreeNode * node = this;
+  for ( const auto & attrName : path )
+    {
+      try
+        {
+          node = &node->children.at( attrName );
+        }
+      catch ( const std::out_of_range & err )
+        {
+          return SR_DEFAULT;
+        }
+    }
+  return node->rule;
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+std::optional<bool>
+RulesTreeNode::applyRules( const AttrPath & path ) const
+{
+  auto rule = this->getRule( path );
+  /* Perform lookup in parents if necessary. */
+  if ( rule == SR_DEFAULT )
+    {
+      AttrPath pathCopy = path;
+      do {
+          pathCopy.pop_back();
+          rule = this->getRule( pathCopy );
+        }
+      while ( ( rule == SR_DEFAULT ) && ( ! pathCopy.empty() ) );
+    }
+
+  switch ( rule )
+    {
+      case SR_ALLOW_PACKAGE: return true;
+      case SR_DISALLOW_PACKAGE: return false;
+      case SR_ALLOW_RECURSIVE: return true;
+      case SR_DISALLOW_RECURSIVE: return false;
+      case SR_DEFAULT: return std::nullopt;
+      default:
+        throw PkgDbException( "encountered unexpected rule `"
+                              + scrapeRuleToString( rule ) + '\'' );
+    }
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+void
+from_json( const nlohmann::json & jfrom, RulesTreeNode & rules )
+{
+  ScrapeRulesRaw raw = jfrom;
+  rules              = static_cast<RulesTreeNode>( raw );
```

Comment:
nit/personal preference
```suggestion
  rules              = RulesTreeNode( raw );
```